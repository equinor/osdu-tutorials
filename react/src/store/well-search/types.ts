import { FindWellsResponse, FindWellboresResponse } from 'api/well-search.api';

export const FIND_WELLS_BY_NAME_START = 'FIND_WELLS_BY_NAME_START';
export const FIND_WELLS_BY_NAME_SUCCESS = 'FIND_WELLS_BY_NAME_SUCCESS';
export const FIND_WELLS_BY_NAME_FAIL = 'FIND_WELLS_BY_NAME_FAIL';

export const FIND_WELLBORES_START = 'FIND_WELLBORES_START';
export const FIND_WELLBORES_SUCCESS = 'FIND_WELLBORES_SUCCESS';
export const FIND_WELLBORES_FAIL = 'FIND_WELLBORES_FAIL';

export interface FindWellsStartAction {
  type: typeof FIND_WELLS_BY_NAME_START;
  payload: string;
}

export interface FindWellsSuccessAction {
  type: typeof FIND_WELLS_BY_NAME_SUCCESS;
  payload: FindWellsResponse;
}

export interface FindWellsFailAction {
  type: typeof FIND_WELLS_BY_NAME_FAIL;
  payload: Error;
}

export interface FindWellboresStartAction {
  type: typeof FIND_WELLBORES_START;
  payload: string;
}

export interface FindWellboresSuccessAction {
  type: typeof FIND_WELLBORES_SUCCESS;
  payload: {
    wellId: string;
    result: FindWellboresResponse;
  }
}

export interface FindwellboresFailAction {
  type: typeof FIND_WELLBORES_FAIL;
  payload: {
    err: Error;
    wellId: string;
  }
}

export type WellSearchActionTypes =
  | FindWellsStartAction
  | FindWellsSuccessAction
  | FindWellsFailAction
  | FindWellboresStartAction
  | FindWellboresSuccessAction
  | FindwellboresFailAction;
