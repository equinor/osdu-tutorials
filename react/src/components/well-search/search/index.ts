export { ConnectedSearch as Search } from './ConnectedSearch';
