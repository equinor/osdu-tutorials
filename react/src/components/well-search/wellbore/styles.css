.wellbore {
    display: flex;
    justify-content: space-between;
    align-items: center;
    text-indent: 0;
    margin: 0;
    padding-left: 26px;
    font-size: 15px;
    height: 30px;
    letter-spacing: 1px;
}

.wellbore__load-mark {
    position: absolute;
    left: -20px;
    display: flex;
    justify-content: center;
    align-items: center;
    width: 10px;
    height: 100%;
}

.wellbore__control {
    text-align: center;
    width: 85px;
}

.wellbore__action {
    position: relative;
    width: 85px;
    height: 20px;
    line-height: 20px;
    border: none;
    border-radius: 4px;
    text-align: center;
    text-transform: uppercase;
    background-color: #4285f4;
    box-shadow: 0px 5.5px 5px rgba(0, 0, 0, 0.24), 0px 9px 18px rgba(0, 0, 0, 0.18);
    color: white;
    font-size: 12px;
    outline: none;
    transition: box-shadow 0.3s;
    cursor: pointer;
}

.wellbore__action:hover,
.wellbore__action:focus {
    box-shadow: 0 2px 4px -1px rgba(0, 0, 0, 0.2), 0 4px 5px 0 rgba(0, 0, 0, 0.14),
    0 1px 10px 0 rgba(0, 0, 0, 0.12);
}

.wellbore__action:active {
    box-shadow: 0 0 11px rgba(33, 33, 33, 0.2);
}

.wellbore__action--visualize {
    background-color: #4285f4;
}

.wellbore__action--hide {
    background-color: #c2c2c2;
}

.wellbore__action--error {
    background-color: #ff8b8b;
}

.wellbore__action-tooltip {
    display: none;
    position: absolute;
    bottom: 100%;
    left: 0%;
    margin-bottom: 10px;

    text-transform: none;
    letter-spacing: normal;
}

.wellbore__action:hover .wellbore__action-tooltip {
    display: block;
}
