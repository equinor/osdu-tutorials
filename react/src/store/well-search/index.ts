export { findWellsByNameAction, findWellboresAction } from './actions';
export * from './reducer';
