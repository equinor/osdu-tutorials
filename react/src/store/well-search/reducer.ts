import {
  WellSearchActionTypes,
  FIND_WELLS_BY_NAME_START,
  FIND_WELLS_BY_NAME_FAIL,
  FIND_WELLS_BY_NAME_SUCCESS,
} from './types';
import { GeoJSON, LatLng } from 'leaflet';

export interface WellSearchState {
  /** initial or last searched name. */
  searchName: string;

  /** well search api request progress flag */
  areWellsSearching: boolean;

  /** proof that a search request was made */
  areWellsSearched: boolean;

  /** any error, occured during a search request */
  searchError?: Error;

  /** a resulted list of wells after a successful rearch request */
  foundWells: WellSearchResponse[];
}

const initialState: WellSearchState = {
  searchName: 'BIR',
  areWellsSearching: false,
  areWellsSearched: false,
  foundWells: [],
  searchError: undefined,
};

export interface WellSearchResponse {
  resourceId: string;

  /** a geo location, will be useful to reveal the well on the map */
  location: LatLng;
}

export const wellSearchReducer = (
  state: WellSearchState = initialState,
  action: WellSearchActionTypes
): WellSearchState => {
  switch (action.type) {
    case FIND_WELLS_BY_NAME_START:
      return {
        ...state,
        searchName: action.payload,
        areWellsSearching: true,
        areWellsSearched: false,
        searchError: undefined,
      };
    case FIND_WELLS_BY_NAME_SUCCESS:
      return {
        ...state,
        areWellsSearching: false,
        areWellsSearched: true,
        foundWells: action.payload.results.map(
          (well): WellSearchResponse => ({
            resourceId: well.id,
            location: new LatLng(
              well.data['SpatialLocation.Wgs84Coordinates'].geometries[0].coordinates[1],
              well.data['SpatialLocation.Wgs84Coordinates'].geometries[0].coordinates[0]
            )
          })
        ),
      };
    case FIND_WELLS_BY_NAME_FAIL:
      return {
        ...state,
        areWellsSearching: false,
        areWellsSearched: true,
        searchError: action.payload,
      };
    default:
      return state;
  }
};
