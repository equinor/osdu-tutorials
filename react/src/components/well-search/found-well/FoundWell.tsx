import React from 'react';
import './styles.css';
import { WellSearchResponse } from 'store/well-search';

export interface FoundWellProps {
  /** a well model to be represented by the component */
  well: WellSearchResponse;
}

export function FoundWell({
  well,
}: FoundWellProps) {
  const markClass = ['well__open-mark'].concat('well__open-mark--opened');

  return (
    <div className="well">
      <div className="well__label-container">
        <label className="well__label">
          {/* despite a special responsive icon, the whole name is clickable */}
          {/* not to force a user into a pixel-hunting */}

          <div className="well__label-mark">
            {/* an 'agle' icon, representing a drop-down behavior */}
            {/* it will be replaced with a load icon for the wellbores fetching process */}
            {(
                <span className={markClass.join(' ')} />
            )}
          </div>

          <span>{well.resourceId}</span>
        </label>
      </div>
    </div>
  );
}
