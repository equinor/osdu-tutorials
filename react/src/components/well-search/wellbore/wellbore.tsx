import React from 'react';
import './styles.css';
import { Wellbore as WellboreModel } from 'store/well-search';
import { TrajectoryToDraw } from 'store/trajectory';
import { Loader } from 'components/shared';
import { TooltipPane } from 'components/shared/tooltip-pane';

export interface WellboreProps {
  /** a wellbore model to be represented by the component */
  wellbore: WellboreModel;

  /** trajectories to draw a visualize button in all its forms */
  selectedTrajectories: TrajectoryToDraw[];

  /** a toggle of a trajectory visualization */
  onSelected: (srn: string, selected: boolean) => void;
}

export function Wellbore({ wellbore, selectedTrajectories, onSelected }: WellboreProps) {
  console.log("wellog", wellbore)
  const trajectory = selectedTrajectories.find(st => st.wellboreId === wellbore.id);
  const toggleTrajectorySelected = () => {
    onSelected(wellbore.id, trajectory !== undefined);
  };

  return (
    <li className="wellbore">
      <span>{wellbore.id}</span>
      {/* an action-button/satus to the right to interact with trajectory 3D visualization feature */}
      <div className="wellbore__control">{getControl(trajectory, toggleTrajectorySelected)}</div>
    </li>
  );
}

/**
 * The so-called 'visualize-button' in fact has multiple variations
 * and they are groupped is this method not to bloat main component template
 * *maybe it's a sign to move the code to a separate file
 * @param trajectory a selected trajectory model - we need its state descriptors
 * @param toggleTrajectorySelected add/remove from a list to visualize the trajectory
 */
function getControl(
  trajectory: TrajectoryToDraw | undefined,
  toggleTrajectorySelected: () => void
) {
  if (!trajectory) {
    // a visualize button itself
    return (
      <div
        className="wellbore__action wellbore__action--visualize"
        onClick={toggleTrajectorySelected}>
        Visualize
      </div>
    );
  }

  if (trajectory.isLoading) {
    return <Loader size={8} />;
  }

  if (trajectory.loadError !== undefined) {
    // an error "state-button" will be displayed and a description available in a tooltip
    // an attached toggle action will allow to re-run fetch process
    return (
      <div className="wellbore__action wellbore__action--error" onClick={toggleTrajectorySelected}>
        Error
        <TooltipPane className="wellbore__action-tooltip" doApplyStylesForText={true}>
          {trajectory.loadError.message}
        </TooltipPane>
      </div>
    );
  }

  if (trajectory.isLoaded) {
    // a success "state-button", activation will hide trajectory's visualization
    return (
      <div className="wellbore__action wellbore__action--hide" onClick={toggleTrajectorySelected}>
        Hide
      </div>
    );
  }
  return null;
}
