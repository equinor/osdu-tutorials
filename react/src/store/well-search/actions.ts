import {AppState} from 'store';
import {ThunkAction} from 'redux-thunk';
import {
  FIND_WELLBORES_FAIL,
  FIND_WELLBORES_START,
  FIND_WELLBORES_SUCCESS,
  FIND_WELLS_BY_NAME_FAIL,
  FIND_WELLS_BY_NAME_START,
  FIND_WELLS_BY_NAME_SUCCESS,
  WellSearchActionTypes,
} from './types';
import {FindWellboresResponse, findWellsByName, FindWellsResponse, findWellbores} from 'api/well-search.api';

type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppState, null, WellSearchActionTypes>;

export function findWellsByNameStartAction(wellName: string): WellSearchActionTypes {
  return {
    type: FIND_WELLS_BY_NAME_START,
    payload: wellName,
  };
}

export function findWellsByNameSuccessAction(response: FindWellsResponse): WellSearchActionTypes {
  return {
    type: FIND_WELLS_BY_NAME_SUCCESS,
    payload: response,
  };
}

export function findWellsByNameFailAction(err: Error): WellSearchActionTypes {
  return {
    type: FIND_WELLS_BY_NAME_FAIL,
    payload: err,
  };
}

export function findWellboresStartAction(wellId: string): WellSearchActionTypes {
  return {
    type: FIND_WELLBORES_START,
    payload: wellId,
  };
}

export function findWellboresSuccessAction(
    wellId: string,
    response: FindWellboresResponse
): WellSearchActionTypes {
  return {
    type: FIND_WELLBORES_SUCCESS,
    payload: {
      wellId: wellId,
      result: response,
    }
  };
}

export function findWellboresFailAction(err: Error, wellId: string): WellSearchActionTypes {
  return {
    type: FIND_WELLBORES_FAIL,
    payload: {err, wellId: wellId}
  };
}


/**
 * fetch from the osdu backend a list of wells, matching a querried name
 * @param wellName
 */
export const findWellsByNameAction = (wellName: string): AppThunk => dispatch => {
  dispatch(findWellsByNameStartAction(wellName));

  return findWellsByName(wellName)
    .then(data => dispatch(findWellsByNameSuccessAction(data)))
    .catch(err => dispatch(findWellsByNameFailAction(err)));
};

/**
 * fetch from the backend a list of wellbores for a given well
 * @param wellId
 */
export const findWellboresAction = (wellId: string): AppThunk => dispatch => {
  dispatch(findWellboresStartAction(wellId));

  return findWellbores(wellId)
      .then(data => dispatch(findWellboresSuccessAction(wellId, data)))
      .catch(err => dispatch(findWellboresFailAction(err, wellId)));
};
