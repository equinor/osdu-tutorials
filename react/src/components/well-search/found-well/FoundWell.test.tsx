import React from 'react';
import { shallow } from 'enzyme';
import { FoundWell, FoundWellProps } from './FoundWell';
import { wellSearchWellsFindSuccessState } from 'test-data/well-search';
import { Wellbore } from '../wellbore';

describe('FoundWell', () => {
  it('should render self', () => {
    const props: FoundWellProps = {
      selectedTrajectories: [],
      well: wellSearchWellsFindSuccessState.foundWells[0],
      onLoadWellbores: jest.fn(),
      onFetchTrajectory: jest.fn(),
      onUnselectTrajectory: jest.fn(),
    };
    const wrapper = shallow(<FoundWell {...props} />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Wellbore).length).toBe(props.well.wellbores.length);
  });
});
