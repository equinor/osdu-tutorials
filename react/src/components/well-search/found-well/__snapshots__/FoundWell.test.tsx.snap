// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FoundWell should render self 1`] = `
<div
  className="well"
>
  <div
    className="well__label-container"
  >
    <label
      className="well__label"
      onClick={[Function]}
    >
      <div
        className="well__label-mark"
      >
        <span
          className="well__open-mark"
        />
      </div>
      <span>
        WellName1
      </span>
    </label>
  </div>
  <ul
    className="well__trajectories-list"
  />
</div>
`;
